File: 0876-middle-of-the-linked-list/0876-middle-of-the-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode *tortoise = head;
        ListNode *hare = head->next;
        if (hare == nullptr) return tortoise;
        while (hare->next != nullptr && hare->next->next != nullptr) {
            hare = hare->next->next;
            tortoise = tortoise->next;
        }
        return tortoise->next;
    }
};

File: 0876-middle-of-the-linked-list/0876-middle-of-the-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        if (head.next == null) return head;
        ListNode baby = head;
        ListNode giant = head.next;
        while (giant != null) {
            if (giant.next == null || giant.next.next == null) break;
            giant = giant.next;
            giant = giant.next;
            baby = baby.next;
        }
        return baby.next;
    }
}